### PR TITLE
✨ Bento Selector: keyboard-select-mode feature

### PR DESCRIPTION
**Pull request by caroqliu**
Partial for #28282. This PR introduces the following changes:
- `keyboard-select-mode` attribute on `amp-selector` with the following possible configurations:
  - `none`: `Enter` and `Space` selects an `Option` (Default)
  - `focus`: `Enter` and `Space` selects an `Option`, Arrow keys `Left`, `Right`, `Down`, and `Up` navigate focus in DOM order between options, including looping:
15fa9543-3165-4f0c-ad46-b04173d4824d
  - `select`: Arrow keys `Left`, `Right`, `Down`, and `Up` navigate selected state on the target element in DOM order between options, including looping.
6639fd69-71a2-4706-92e4-122c0cea37d5
- `Enter` and `Space` selects an `Option` in Preact mode, but no `keyboardSelectMode` prop is used as this requires a defined order of option children. Omitting `keyboardSelectMode` should be fine in this layer, as it is primarily a prop for enhanced configuration of keyboard navigability in the component, however by default the options are already keyboard navigable. If Preact developers hope to add additional keyboard events to change selection state, this is supported via the exposed `toggle` action.

### CODE REVIEW

**Comment by dvoytenko** [comment]
This looks good, but the concern over code-split becomes even higher here. Wouldn't we want some of these keys implemented on Preact level? And if so, would asking a developer to supply ordering index to enable this be a big ask? I'm asking this because our cost is growing for this split and we might be losing some features from the Preact side. But perhaps asking the developer to supply an additional data is not such a big problem?

**Comment by caroqliu** [comment]
> This looks good, but the concern over code-split becomes even higher here. Wouldn't we want some of these keys implemented on Preact level? And if so, would asking a developer to supply ordering index to enable this be a big ask? I'm asking this because our cost is growing for this split and we might be losing some features from the Preact side. But perhaps asking the developer to supply an additional data is not such a big problem?

@dvoytenko  Added a couple more commits to implement and discuss using `order` in the `Option` component. This includes:
- moving `selectBy` to Preact layer imperative API, which is noop if no `Option` has a non-zero or non-positive integer `order` prop
- Reversing `Selector` and `SelectorShim` as the `amp-selector['Component']`/`"as"` prop, so that the `onKeyDown` defined in `Selector` can be registered on the `SelectorShim`'s `shimDomElement`.
- Add `keyboardSelectMode` to the Storybook samples
- Restore the unit tests for `selectBy` that were removed in #31068

I agree this simplifies code divergence and maintenance between the layers over time. The optional use of `order` prop is smooth as well, but I am leaning towards renaming it as `index` since (1) `order` may be confused with the CSS property, which we do not read, and (2) `index` helps imply the concept of valid values being `>=0` integers.

WDYT?

**Review by dvoytenko — CHANGES_REQUESTED** [review]
Yes. I think this is a better structure. And I agree - we should rename `order` to `index`.

File: extensions/amp-selector/1.0/amp-selector.js [hunk 1]
```diff
@@ -88,23 +82,30 @@ class AmpSelector extends PreactBaseElement {
...
-      this.mutateProps({children});
+      this.mutateProps({children, options});
     });
     mu.observe(element, {
-      attributeFilter: ['option', 'selected', 'disabled'],
+      attributeFilter: [
+        'option',
+        'selected',
+        'disabled',
+        'keyboard-select-mode',
```
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-selector/1.0/amp-selector.js:92]: Good catch, thanks.

**Review by dvoytenko — APPROVED** [review]

**Pull request merged**